### DISCUSSION TITLE
How to create a vcproj (or equiv) in Visual Studio from our CMake build?

### DISCUSSION THREAD

**Discussion by PhilipOakley** [git-for-windows/git#3176] [Q&A]
The new `CMake` method for building Git for Windows (GfW) in Visual Studio (2019) works nicely, but unfortunately many support tools are expecting there to be (I suspect) `vcproj` and maybe `sln` files also present so those extension tools can do their work.

I'm trying to get [Sourcetrail](https://github.com/CoatiSoftware/Sourcetrail) integrated with the current release to help browse the git code. It had worked previously (year ago+, using `vs/master` branch, VS 2015, Sourcetrail v2019). 

I raised an issue over there https://github.com/CoatiSoftware/Sourcetrail/issues/1173 and tried their suggestion of using a Clang helper tool, but that also doesn't think we have created a project either.

Is there a way to get either our `CMake` invocation, or a Visual Studio command to generate the old style `vcproj` files so that extension tools that expect them continue to work?

**Comment by dscho** [comment]
Yes, you can follow [the advice in `CMakeLists.txt`](https://github.com/git-for-windows/git/blob/67e748cc61a2a389d04feedca7ae9cc63f2354b4/contrib/buildsystems/CMakeLists.txt#L19-L21):

> ```
>     mkdir -p contrib/buildsystems/out
>     cd contrib/buildsystems/out
>     cmake ../ -DCMAKE_BUILD_TYPE=Release
> ```

This requires [CMake](https://cmake.org/) to be installed separately, and will result in a `git.sln` in `contrib/buildsystems/out/`.

**Reply by PhilipOakley** [reply]
> This requires [CMake](https://cmake.org/) to be installed separately, and will result in a `git.sln`

@dscho Many thanks. Those two extra points do clarify a few things that were missing.

Background [long explanation]: I've never used `CMake` before, so my understanding of it's operation had been inferred  from the discussions about VS now being able to use it, and is simplifying the maintenance of 'the project'. 

Hence I expected that if I open the Git for Windows project folder in a new install of Visual Studio (2019 community ed), then VS would be able to build the 'project', which it did, and I could run the tests (from the menu options). All good so far, for me.

I now tried to integrate with [Sourcetrail](https://github.com/CoatiSoftware/Sourcetrail) (hopes of getting back to the 4GB mega problem). But that failed. I thought I had a 'project', but in some peculiar way I didn't have a 'project'. Riddle me confused.

I had see the instructions in [`CMakeLists.txt`](https://github.com/git-for-windows/git/blob/67e748cc61a2a389d04feedca7ae9cc63f2354b4/contrib/buildsystems/CMakeLists.txt#L19-L21), as well as the final [`vs/main` commit](https://github.com/git-for-windows/git/tree/vs/main), but read that as being an alternative way to build (compile) the project and its executables that bypassed any VS dependency. However I was using VS, so I thought VS would be doing all that (it does..) and that the 'project' would integrate directly with all the other historic VS tools (but it didn't..). 

All this shows that I wasn't aware of that tool dependency on explicitly generating a `.sln` file (and likely the associated `.vcproj` files - not checked yet), and that VS itself now has, for the purpose of some tool integrations, two types of 'project' - (A) those from a `.sln`/`.vcproj` project, and (B) those from a CMake build, and that many tools (i.e. the one I was trying) needed the type A project. And worse, even though VS could build via a `CMakeLists.txt` it doesn't, won't, can't (not sure which) generate those associated `.sln`/`.vcproj` files in an obvious manner - which is where I was stuck.

I hadn't realised are that even though I have VS installed, I need those extra (non-VS) steps to generate the `.sln` project needed by VS too integrate with many other VS extensions that have a hidden dependency of its existence. 

I hope to check my understanding above in the next few days (Lockdown has just been released here, so getting out first..) to check that I can use the Sourcetrail extension with VS with the Git-for-Windows project. Assuming it all fits I'll see if I can add a few words to the `CMakeLists.txt` file and the final `vs/main` and `vs/master` commits for others in the same situation.

I hope that helps clarify how I got into that hole.

**Reply by PhilipOakley** [reply]
> I hope to check my understanding above in the next few days (Lockdown has just been released here, so getting out first..) to check that I can use the Sourcetrail extension with VS with the Git-for-Windows project.

Progress has been 2 steps forward but at least 1 step back.. Some notes.

I started a terminal in VS (not something I've needed in the past) and checked out what VS thought it had.
```
PS C:\git-sdk-64\usr\src\git> cmake --version
cmake version 3.19.20122902-MSVC_2
```

I tried the build commands from the terminal, which failed (2 errors), from the VS terminal. The `/out` directory was already present. 
```
PS C:\git-sdk-64\usr\src\git\contrib\buildsystems\out> cmake ../ -DCMAKE_BUILD_TYPE=Release
-- Building for: Visual Studio 16 2019
CMake Warning at C:/git-sdk-64/usr/src/git/compat/vcbuild/vcpkg/scripts/buildsystems/vcpkg.cmake:104 (message):
  There are no libraries installed for the Vcpkg triplet x86-windows.
Call Stack (most recent call first):
  C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.19/Modules/CMakeDetermineSystem.cmake:104 (include)
  CMakeLists.txt:91 (project)


-- The C compiler identification is MSVC 19.28.29913.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
  Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR)
Call Stack (most recent call first):
  C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:582 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.19/Modules/FindZLIB.cmake:115 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  C:/git-sdk-64/usr/src/git/compat/vcbuild/vcpkg/scripts/buildsystems/vcpkg.cmake:263 (_find_package)
  CMakeLists.txt:140 (find_package)
```

I then tried the CMake install - The default install doesn't add `cmake` to the path. I re-installed with it added to path.

Running in a CMD window:
```
C:\git-sdk-64\usr\src\git\contrib\buildsystems\out>cmake --version
cmake version 3.20.1
```

Running the build (in CMD) with that version gave the same two errors ("no libraries installed for the Vcpkg triplet x86-windows" and "Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR)" ).
```
>cmake ../ -DCMAKE_BUILD_TYPE=Release
CMake Warning at C:/git-sdk-64/usr/src/git/compat/vcbuild/vcpkg/scripts/buildsystems/vcpkg.cmake:104 (message):
  There are no libraries installed for the Vcpkg triplet x86-windows.
Call Stack (most recent call first):
  out/CMakeFiles/3.20.1/CMakeSystem.cmake:6 (include)
  CMakeLists.txt:91 (project)


CMake Error at C:/Program Files/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR)
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/CMake/share/cmake-3.20/Modules/FindZLIB.cmake:120 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  C:/git-sdk-64/usr/src/git/compat/vcbuild/vcpkg/scripts/buildsystems/vcpkg.cmake:263 (_find_package)
  CMakeLists.txt:140 (find_package)
```

Web search for the messages suggested some reasons for these, e.g. :
[Triplet ](https://github.com/microsoft/vcpkg/issues/9339) "can "fix" this by changing the VCPKG_TARGET_TRIPLET to `x64-windows` " ?
[Zlib](https://stackoverflow.com/questions/24808150/how-to-point-cmake-to-zlib-include-path) "guess is that it's in a non-standard location" - apparently some need to mark Zlib as 'Advanced'? and tell it where we have it (I have it at C:\git-sdk-64\usr\src\git\compat\vcbuild\vcpkg\buildtrees\zlib)

Anyway, it's some progress.

**Reply by dscho** [reply]
If you built the dependencies with a recent `vcpkg` version, the `CMake` definition should pick `zlib` up automatically. (Did you delete `compat/vcbuild/vcpkg` and re-ran `compat/vcbuild/vcpkg_install.bat`?)

**Reply by PhilipOakley** [reply]
> If you built the dependencies with a recent `vcpkg` version, the `CMake` definition should pick `zlib` up automatically. 
> (Did you delete `compat/vcbuild/vcpkg` and re-ran `compat/vcbuild/vcpkg_install.bat`?)

It had been a while since I re-installed the `vcpkg` directory. 
Given that the Visual Studio [menu] build had worked without error (just the usually warnings about signed/unsigned etc., which I inspected), I was expecting that the command line builds would work just the same.

I did remember that I'd need to review the `vcpkg` updates, but time was against me - thank you for confirming the command line instructions. 

I hope to follow up today so that we can cover these setup issues in the instructions (age brings too many things to remember, or where one made the notes 😉 )

**Reply by PhilipOakley** [reply]
@dscho After deleting vcpkg and reinstalling it, and trying variants, it is still failing. 

It (the missing ZLib) is a very common error symptom on the web, with few 'solutions', many of which are similar to "just to get the gnu zlib and that will be found automatically", or "use the cmake-gui and point directly at your zlib", neither of which are a proper solution. 

I added some information messages to the script to show the paths that CMake has, at the point it's attempting finding Zlib.
```
message("CMAKE_SOURCE_DIR: ${CMAKE_SOURCE_DIR}")
message("VCPKG_DIR: ${VCPKG_DIR}")
message("CMAKE_PREFIX_PATH: ${CMAKE_PREFIX_PATH}")
```
which resulted in these three path variable, the last being a three item list, with linebreaks for readability
```
CMAKE_SOURCE_DIR: C:/git-sdk-64/usr/src/git/contrib/buildsystems/../..

VCPKG_DIR: C:/git-sdk-64/usr/src/git/contrib/buildsystems/../../compat/vcbuild/vcpkg

CMAKE_PREFIX_PATH: 
C:/git-sdk-64/usr/src/git/contrib/buildsystems/../../compat/vcbuild/vcpkg/installed/;
C:/git-sdk-64/usr/src/git/compat/vcbuild/vcpkg/installed/x86-windows;
C:/git-sdk-64/usr/src/git/compat/vcbuild/vcpkg/installed/x86-windows/debug
```
I'm suspicious that these `contrib/buildsystems/../../` may have confused `cmake`, but it may also be an internal options to `findZLib` that require precise paths that is the issue.

The real paths I have are:
`C:\git-sdk-64\usr\src\git\compat\vcbuild\vcpkg\installed\x64-windows` (directory)

`C:\git-sdk-64\usr\src\git\compat\vcbuild\vcpkg\installed\vcpkg\info\zlib_1.2.11_x64-windows.list` (file)
with .list contents:
```
x64-windows/
x64-windows/bin/
x64-windows/bin/zlib.pdb
x64-windows/bin/zlib1.dll
x64-windows/debug/
x64-windows/debug/bin/
x64-windows/debug/bin/zlibd.pdb
x64-windows/debug/bin/zlibd1.dll
x64-windows/debug/lib/
x64-windows/debug/lib/pkgconfig/
x64-windows/debug/lib/pkgconfig/zlib.pc
x64-windows/debug/lib/zlibd.lib
x64-windows/include/
x64-windows/include/zconf.h
x64-windows/include/zlib.h
x64-windows/lib/
x64-windows/lib/pkgconfig/
x64-windows/lib/pkgconfig/zlib.pc
x64-windows/lib/zlib.lib
x64-windows/share/
x64-windows/share/zlib/
x64-windows/share/zlib/copyright
x64-windows/share/zlib/usage
x64-windows/share/zlib/vcpkg_abi_info.txt
```
These all appear to point correctly.

A few zlib threads, e.g. https://cmake.org/pipermail/cmake/2017-August/066161.html https://cmake.org/cmake/help/latest/module/FindZLIB.html#hints (variously imply giving a path direct to zlib) 

Any suggestions for further debugging?

**Reply by PhilipOakley** [reply]
@dscho, some success

I added `set(ZLIB_ROOT "${CMAKE_SOURCE_DIR}/compat/vcbuild/vcpkg/installed/x64-windows")` just before the `find_package(ZLIB REQUIRED)`, and this has allowed Zlib to be found,
 and the `.sln` to be generated,
 and Visual Studio builds that,
 and the Sourcetrail Extension to detect the solution!

Some progress, but still need to see how to clear the `/../..` path element (and confirm it was the cause..).

**Reply by PhilipOakley** [reply]
Found it: NOT EXISTS ${VCPKG_ARCH}

```
	if(NOT EXISTS ${VCPKG_ARCH})
	set(VCPKG_ARCH "x64-windows") # default from vcpkg_install.bat
	endif()
```
Q: can vcpkg tell cmake what the installed arch is?

**Reply by dscho** [reply]
Apparently this is not something CMake can tell. If you need to configure it for anything but `x64-windows`, you need to set it on the command line (see `.github/workflows/main.yml` and/or `.github/workflows/git-artifacts.yml` for details).

**Reply by PhilipOakley** [reply]
> If you need to configure it for anything but `x64-windows`, you need to set it on the command line

At the moment it looks like I even need to set `VCPKG_ARCH` for `x64-windows` on the command line as well. However if the `vcpkg` isn't installed, and `VCPKG_ARCH` isn't set, then the `vcpkg_install.bat` will assume that `x64-windows` is the one to install (there's a line in the `.bat` file to check & select it), but `VCPKG_ARCH` still isn't set.

I do see that the `VCPKG_TARGET_ARCH` (note the `TARGET` in the middle) is set to `x64-windows` when I use the `CMake-gui` and its Advance listing, but I'm not sure when `CMake` picks that up, and whether that's the variable I could use to determine which vcpkg target has already been installed. 

If CMake checks the vcpkg install when it starts then that may be a 'deliberate' option to chose, or I can just simply add the `if()` lines, and perhaps a user message, to get past that problem. 

I see that the `.yml` files (thanks for the pointer) use the `vcpkg_copy_dlls.bat` rather than the `vcpkg_install.bat` that is used for the simple `cmake` invocation, which maybe where some of the confusion come from about the `x64-windows` defaults.

**Reply by PhilipOakley** [reply]
to correct myself, it's not `VCPKG_TARGET_ARCH` but `VCPKG_TARGET_TRIPLET` that is set to `x64-windows` when watching the cmake-gui.

**Comment by PhilipOakley** [comment]
The discussion continued in #3306, where the CMakeLists.txt file is being updated such that opening the git **folder** in Visual Studio(VS), is sufficient for VS to properly load the project and allows the project to be built via the 'build' menu.
